Improve Panache projection with Hibernate ORM and Reactive

DavideD opened the pull request:
Fixes #26284

It makes it possible to use projection with HQL queries with a select clause.

@loicmathieu this seems to work.

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -82,6 +82,23 @@ public <T> CommonPanacheQueryImpl<T> project(Class<T> type) {
+            throw new PanacheQueryException("Unable to perform a projection on a 'select new' query");
+        }
+
+        if (query.toLowerCase().startsWith("select ")) {
+            int endSelect = query.indexOf(" from ");
+            String from = query.substring(endSelect);
+            StringBuilder newQuery = new StringBuilder("SELECT new ")
+                    .append(type.getName())
+                    .append(" (")
+                    .append(query.substring("SELECT ".length(), endSelect))
+                    .append(")")
+                    .append(from);
+
+            return new CommonPanacheQueryImpl<>(this, newQuery.toString(), "select count(*) " + from);
+        }
```
I wonder if it'll cover a lot of use case of not.

I saw your test and it seems to working fine for some usage but we already have some request to support query that returns arbitrary types, `SELECT new` will not work for basic types like primitive or String. I wonder if we shouldn't use `jpaQuery.setResultClass()` instead, it will be a more complex change but will be more flexible.

DavideD replied:
What makes you say it won't work for  basic types?
Maybe I'm wrong, but `.setResultClass` doesn't transform the result, it's used to specify what the return type is going to be.

I wanted to write more tests. After spending so much time trying to make the test suite work and fail, I surrendered.

loicmathieu replied:
You're right, it should be tested.

@DavideD what causes issue to run the Hibernate with Panache integration-test ? It uses and embedded database so it's normally working fine.

DavideD replied:
I've added some test cases for when the select contains constants or other functions. It seems to work as expected.
I think we can merge this.

DavideD committed:
[#26284] Support .project for queries with select clause

User can now write an HQL with a select clause and project
will work. It will run a valid HQL that converts each tuple
in the result to selected type, as long as a valid constructor
exists.

DavideD committed:
[#26284] Tests for .project with HQL queries having 'select'

loicmathieu reviewed (commented):
LGTM, I was planning to ask for example with agregated query but you had the same idea and add them already ;)

Can you please update documentation / JavaDoc with a sentence explaing that if the query contains a select clause we do our best to transform it to a constructor expression, else we will build the query based on the constructor parameter list.

I would also like to have the opinion of @FroMage on this one now that the impl is matured and works.

DavideD commented:
I don't mind updating the documentation, but I would like to make sure that we agree on this approach and we are going to merge it before updating the docs.

DavideD committed:
[#26284] Support select-distinct

DavideD committed:
[#26284] Test select-distinct with .project

FroMage commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -82,6 +82,37 @@ public <T> CommonPanacheQueryImpl<T> project(Class<T> type) {
+        // using the fields in the select clause:
+        // Initial query: select e.field1, e.field2 from EntityClass e
+        // New query: SELECT new org.acme.ProjectionClass(e.field1, e.field2) from EntityClass e
+        if (query.trim().toLowerCase().startsWith("select ")) {
```
Davide :) Please :) My eyes are bleeding from all those `query.trim().toLowerCase()` calls :)

DavideD replied:
Is this better? https://github.com/quarkusio/quarkus/pull/26510/commits/966f6af57bd943905851d7e52bec1d562afff791

FroMage reviewed (changes requested):
This is great, IMO, and was missing. Can you do the same for HR and add documentation?

DavideD committed:
[#26284] Minor refactoring

* Reduce number of `trim()` and `toLowerCase()` calls
* Remove unecessary calls to `.length()`

DavideD commented:
I've added an example in the guides and refactor the code to remove some of the trims.
Let me know.

DavideD commented:
Ah nevermind, I need to add the feature to reactive. I was expecting this logic to be shared

DavideD committed:
[#26284] Add support for Hibernate Reactive

Support `.project` with HQL queries with a select clause

DavideD committed:
[#26284] Test .project with Hibernate Reactive Panache

Add tests with HQL queries with a select clause and .project

DavideD committed:
[#26284] Update ORM and Reactive Panache guides

Add example for Panache .project with HQL query with a select clause

DavideD committed:
[#26284] Fix pre-existing typo in comment

DavideD commented:
I've added the missing part for Hibernate Reactive, the guides and refactor the code. Something is failing on CI but I don't think it's related to these changes. It should be good to go

FroMage reviewed (approved):
LGTM

FroMage commented:
@loicmathieu are you OK for me to merge this?

loicmathieu commented:
@FroMage yes, it's OK for me, I follow the discussion and had nothing to say ;)

gsmet commented on docs/src/main/asciidoc/hibernate-orm-panache.adoc:
```diff
@@ -791,6 +792,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+It's also possible to specify a HQL query with a select clause. In this case, the projection class must have a constructor
+matching the values returned by the select clause:
+
+source,java]
```
```suggestion
[source,java]
```

gsmet commented on docs/src/main/asciidoc/hibernate-orm-panache.adoc:
````diff
@@ -791,6 +792,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+It's not possible to have a HQL `select new` query and `.project(Class)` at the same time - you need to pick one approach.
+
+For example, this will fail:
+```
````
Please use a proper Java source block here.

gsmet commented on docs/src/main/asciidoc/hibernate-orm-panache.adoc:
```diff
@@ -791,6 +792,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
 ----
 <1> The `ownerName` DTO constructor's parameter will be loaded from the `owner.name` HQL property.
 
+It's also possible to specify a HQL query with a select clause. In this case, the projection class must have a constructor
```
I think we try to avoid `it's` in docs (not saying we don't have many occurrences but let's try to not introduce new ones).

```suggestion
It is also possible to specify a HQL query with a select clause. In this case, the projection class must have a constructor
```

gsmet commented on docs/src/main/asciidoc/hibernate-orm-panache.adoc:
```diff
@@ -791,6 +792,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+// Only the race and the average weight will be loaded
+PanacheQuery<RaceWeight> query = Person.find("select d.race, AVG(d.weight) from Dog d group by d.race).project(RaceWeight.class);
+----
+<1> Hibernate ORM will use this constructor. When the query has a select clause, it's possible to have multiple constructors.
```
```suggestion
<1> Hibernate ORM will use this constructor. When the query has a select clause, it is possible to have multiple constructors.
```

gsmet commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -674,6 +675,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
 ----
 <1> The `ownerName` DTO constructor's parameter will be loaded from the `owner.name` HQL property.
 
+It's also possible to specify a HQL query with a select clause. In this case, the projection class must have a constructor
```
```suggestion
It is also possible to specify a HQL query with a select clause. In this case, the projection class must have a constructor
```

gsmet commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -674,6 +675,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+It's also possible to specify a HQL query with a select clause. In this case, the projection class must have a constructor
+matching the values returned by the select clause:
+
+source,java]
```
```suggestion
[source,java]
```

gsmet commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -674,6 +675,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+// Only the race and the average weight will be loaded
+PanacheQuery<RaceWeight> query = Person.find("select d.race, AVG(d.weight) from Dog d group by d.race).project(RaceWeight.class);
+----
+<1> Hibernate Reactive will use this constructor. When the query has a select clause, it's possible to have multiple constructors.
```
```suggestion
<1> Hibernate Reactive will use this constructor. When the query has a select clause, it is possible to have multiple constructors.
```

gsmet commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -674,6 +675,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+
+[WARNING]
+====
+It's not possible to have a HQL `select new` query and `.project(Class)` at the same time - you need to pick one approach.
```
```suggestion
It is not possible to have a HQL `select new` query and `.project(Class)` at the same time - you need to pick one approach.
```

gsmet commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
````diff
@@ -674,6 +675,43 @@ PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+It's not possible to have a HQL `select new` query and `.project(Class)` at the same time - you need to pick one approach.
+
+For example, this will fail:
+```
````
Please use a proper Java source block.

gsmet commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -69,6 +69,35 @@ public <T> CommonPanacheQueryImpl<T> project(Class<T> type) {
+        // using the fields in the select clause:
+        // Initial query: select e.field1, e.field2 from EntityClass e
+        // New query: SELECT new org.acme.ProjectionClass(e.field1, e.field2) from EntityClass e
+        if (lowerCasedTrimmedQuery.startsWith("select ")) {
```
Can't say I'm a big fan of all these string manipulations. My guess is that it's going to be brittle and we will get bitten by this at some point but I suppose we will see.

DavideD replied:
I'm not going to write a compiler if that's what you are suggesting

gsmet replied:
I'm not really suggesting that for now. As I said, let's see how it goes.

gsmet reviewed (changes requested):
There are some problems in the documentation. The main one being the missing ``[``'s. Always a good idea to do a doc rendering when introduction new sections that are not trivial.

DavideD committed:
[#26284] Replace "it's" with "it is" in docs

DavideD commented:
I've updated the PR and added a commit that replaces all the `it's` to `it is`: https://github.com/quarkusio/quarkus/pull/26510/commits/38a2a054abcbe95c19e2b4a03376af3927a4e6a0

gsmet commented:
Thanks!
